tracer/src/Datadog.Trace/Agent/StatsAggregator.cs

```diff
@@ -0,0 +1,198 @@
+using System.Threading.Tasks;
+using Datadog.Trace.Configuration;
+using Datadog.Trace.ExtensionMethods;
+using Datadog.Trace.Logging;
+using Datadog.Trace.PlatformHelpers;
+
+namespace Datadog.Trace.Agent
+{
+    internal class StatsAggregator : IDisposable
+    {
+        private const int BufferCount = 2;
+
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor<StatsAggregator>();
+
+        private readonly StatsBuffer[] _buffers;
+
+        private readonly IApi _api;
+
+        private readonly CancellationTokenSource _cancellationTokenSource;
+
+        private readonly TimeSpan _duration;
+
+        private int _currentBuffer;
+
+        public StatsAggregator(IApi api, ImmutableTracerSettings settings)
+             : this(api, settings, TimeSpan.FromSeconds(10))
+        {
+        }
+
+        internal StatsAggregator(IApi api, ImmutableTracerSettings settings, TimeSpan duration)
+        {
+            _api = api;
+            _cancellationTokenSource = new CancellationTokenSource();
+            _duration = duration;
+            _buffers = new StatsBuffer[BufferCount];
+
+            var header = new ClientStatsPayload
+            {
+                Environment = settings.Environment,
+                ServiceName = settings.ServiceName,
+                Version = settings.ServiceVersion,
+                HostName = HostMetadata.Instance.Hostname
+            };
+
+            for (int i = 0; i < _buffers.Length; i++)
+            {
+                _buffers[i] = new(header);
+            }
+
+            if (settings.TracerStatsEnabled)
+            {
+                _ = Task.Run(() => Flush(_cancellationTokenSource.Token), _cancellationTokenSource.Token)
+                    .ContinueWith(t => Log.Error(t.Exception, "Error in StatsAggregator"), TaskContinuationOptions.OnlyOnFaulted);
+            }
+        }
+
+        /// <summary>
+        /// Gets the current buffer.
+        /// StatsBuffer is not thread-safe, this property is not intended to be used outside of the class,
+        /// except for tests.
+        /// </summary>
+        internal StatsBuffer CurrentBuffer => _buffers[_currentBuffer];
+
+        public void Dispose()
+        {
+            _cancellationTokenSource.Cancel();
+        }
+
+        public void Add(params Span[] spans)
+        {
+            AddRange(spans, 0, spans.Length);
+        }
+
+        public void AddRange(Span[] spans, int offset, int count)
+        {
+            // Contention around this lock is expected to be very small:
+            // AddRange is called from the serialization thread, and concurrent serialization
+            // of traces is a rare corner-case (happening only during shutdown).
+            // The Flush thread only acquires the lock long enough to swap the metrics buffer.
+            lock (_buffers)
+            {
+                for (int i = 0; i < count; i++)
+                {
+                    Add(spans[offset + i]);
```

Comment:
I realize this isn't a call to the above `void Add(params Span[] spans)` method but it does look like it could be. I think it would be less confusing if the method was called `AddToBuffer` or something like that.